prov/shm/src/smr_ep.c

```diff
@@ -251,19 +253,37 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct smr_ep_entry *entry,
 
 void smr_progress_cmd(struct smr_ep *ep)
 {
+	struct smr_recv_queue *recv_queue;
+	struct smr_recv_match_attr match_attr;
+	struct dlist_entry *dlist_entry;
 	struct smr_ep_entry *entry;
 	struct smr_cmd *cmd;
 	fi_addr_t addr;
 	int ret = 0;
 
 	fastlock_acquire(&ep->region->lock);
 	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
-	while (!ofi_cirque_isempty(smr_cmd_queue(ep->region)) &&
-	       !ofi_cirque_isempty(ep->rxq)) {
-		
-		entry = ofi_cirque_head(ep->rxq);
+	while (!ofi_cirque_isempty(smr_cmd_queue(ep->region))) {
+
 		cmd = ofi_cirque_head(smr_cmd_queue(ep->region));
 
+		recv_queue = (cmd->hdr.op.flags & FI_TAGGED) ?
+			      &ep->trecv_queue : &ep->recv_queue;
+		if (dlist_empty(&recv_queue->recv_list))
+			break;
+
+		match_attr.addr = FI_ADDR_UNSPEC;
+		match_attr.tag = cmd->hdr.op.tag;
+
+		dlist_entry = dlist_remove_first_match(&recv_queue->recv_list,
+						       recv_queue->match_recv,
+						       &match_attr);
+		if (!dlist_entry) {
+			ret = -FI_EAGAIN;
+			break;
```

Comment:
Here and above we break out of processing if the command at the front of the queue cannot be processed.  See commend above about adding an unexpected queue for tagged messages.  Many apps will not have a matching message ready to receive a tagged message at the time it is sent.  (This is a result of the peers progressing through computations at different rates).

For untagged messages, you may be able to fail at the sender with eagain if no receive buffers have been posted.  You might do this already.